dataflow/src/main/java/org/checkerframework/dataflow/util/IdentityMostlySingleton.java

```diff
@@ -21,21 +22,30 @@ public IdentityMostlySingleton(T value) {
 
     @Override
     @SuppressWarnings("fallthrough")
-    public boolean add(T e) {
+    public boolean add(@FindDistinct T e) {
         switch (state) {
             case EMPTY:
                 state = State.SINGLETON;
                 value = e;
                 return true;
             case SINGLETON:
+                if (value == e) {
+                    return false;
+                }
                 state = State.ANY;
+                // Use ArrayList, but make sure to use reference comparisons.
```

Comment:
In the [other pull request](https://github.com/typetools/checker-framework/pull/3550#issuecomment-668849223), you said that you think using an array list is more efficient.  Can you explain more why you think it is more efficient? `MostlySingleton` uses a HashSet, should we change that one to make it more efficient, too? 

When I was debugging this code, I was very surprised that `set` was in fact an ArrayList.  I wondered why that choice was made. Given that using an ArrayList and using List#contains is not the correct behavior, it made more sense to me to use an identity hash set.  This also matches the implementation match MostlySingleton.